### PR TITLE
api for external relations (using the new logger and validator frameworks)

### PR DESCRIPTION
**Pull request by yrao6597**

### CODE REVIEW

File: Backend/app/controllers/partner.controller.js [hunk 1]
```diff
@@ -1,31 +1,47 @@
...
+      const { name } = req.params;
+      logger.debug(req.params);
+      const partner_info = await Partner.getPartner(name);
+      res.send({ partnerInfo: partner_info });
+    } catch (err) {
+      console.error(err);
+      res.status(500);
+      res.send({ error: err });
     }
+  },
```
Comment (youssefsoli) [inline Backend/app/controllers/partner.controller.js]: Use the new error middleware, where `err` is passed to `next()`. See here https://github.com/uoftblueprint/homes-for-heroes/blob/main/Backend/app/controllers/customer.controller.js for reference.

File: Backend/app/models/partner.model.js [hunk 2]
```diff
@@ -1,38 +1,54 @@
...
+    sql.query(
+      "INSERT INTO partners (org_name, city, village, address, phone) VALUES (?)",
+      [[this.name, this.city, this.village, this.address, this.phone]],
+      function (err, results) {
+        if (err) reject(err);
+        else resolve(results.insertId);
+      }
+    );
+  });
+};
```
Comment (youssefsoli) [inline Backend/app/models/partner.model.js]: Try not to use a two-dimensional array for this application, instead use multiple `?`s and a one-dimensional array.

File: Backend/app/models/partner.model.js [hunk 3]
```diff
@@ -1,38 +1,54 @@
...
+// search for and list a specific partner
+Partner.getPartner = function (partner_name) {
+  return new Promise((resolve, reject) => {
+    sql.query(
+      "SELECT * FROM partners WHERE org_name = ? LIMIT 1",
+      [partner_name],
+      (err, rows) => {
+        if (err) reject(err);
+        else if (!rows[0]) reject(new Error("Partner not found"));
+        else resolve(rows[0]);
```
Comment (youssefsoli) [inline Backend/app/models/partner.model.js]: Try returning a Partner object instead, it makes the API usage easier.

File: Backend/app/models/supporter.model.js [hunk 4]
```diff
@@ -1,37 +1,53 @@
-const sql = require('./db.js');
-const logger = require('../logger');
+const sql = require("./db.js");
+const logger = require("../logger");
```
Comment (youssefsoli) [inline Backend/app/models/supporter.model.js]: It seems like double quotes are being prioritized by the editor, can you try installing an eslint plugin to automatically catch styling issues?

File: Backend/app/models/supporter.model.js [hunk 5]
```diff
@@ -1,37 +1,53 @@
...
-            if (err) reject (err);
-            else resolve(results.insertId);
-        });
-    });
-}
+Supporter.prototype.create = function () {
+  return new Promise((resolve, reject) => {
+    sql.query(
+      "INSERT INTO supporters (name, date_gifted, gift_provided, phone) VALUES (?)",
+      [[this.name, this.date_gifted, this.gift_provided, this.phone]],
```
Comment (youssefsoli) [inline Backend/app/models/supporter.model.js]: Same issue with 2d arrays

File: Backend/app/models/volunteer.model.js [hunk 6]
```diff
@@ -1,37 +1,53 @@
...
-            if (err) reject (err);
-            else resolve(results.insertId);
-        });
-    });
-}
+Volunteer.prototype.create = function () {
+  return new Promise((resolve, reject) => {
+    sql.query(
+      "INSERT INTO volunteers (name, village, date_joined, role, phone) VALUES (?)",
+      [[this.name, this.village, this.date_joined, this.role, this.phone]],
```
Comment (youssefsoli) [inline Backend/app/models/volunteer.model.js]: ditto

**Review by youssefsoli — CHANGES_REQUESTED** [review]

**Comment by yrao6597** [comment]
April 7: fixed code based on the requested changes and added controllers to update external relations info

File: Backend/app/controllers/partner.controller.js [hunk 7]
```diff
@@ -1,31 +1,52 @@
...
+      next(err);
+    }
+  },
+  async updateInfo(req, res, next) {
+    try {
+      logger.debug(req.params);
+      const { partner_id } = req.params;
+      logger.debug(req.body);
+      const updated_id = await Partner.updateInfo(partner_id, req.body);
+      res.send(updated_id);
```
Comment (youssefsoli) [inline Backend/app/controllers/partner.controller.js]: res.json?

File: Backend/app/controllers/supporter.controller.js [hunk 8]
```diff
@@ -1,30 +1,52 @@
...
+      next(err);
+    }
+  },
+  async updateInfo(req, res, next) {
+    try {
+      logger.debug(req.params);
+      const { supporter_id } = req.params;
+      logger.debug(req.body);
+      const updated_id = await Supporter.updateInfo(supporter_id, req.body);
+      res.send(updated_id);
```
Comment (youssefsoli) [inline Backend/app/controllers/supporter.controller.js]: ditto

File: Backend/app/controllers/volunteer.controller.js [hunk 9]
```diff
@@ -1,31 +1,55 @@
...
-            res.status(500);
-            res.send({ error: err });
-        }
+  async updateInfo(req, res, next) {
+    try {
+      logger.debug(req.params);
+      const { volunteer_id } = req.params;
+      logger.debug(req.body);
+      const updated_id = await Volunteer.updateInfo(volunteer_id, req.body);
+      res.send(updated_id);
```
Comment (youssefsoli) [inline Backend/app/controllers/volunteer.controller.js]: ditto

**Review by youssefsoli — CHANGES_REQUESTED** [review]

File: Backend/app/controllers/volunteer.controller.js (hunk 9 above)
Comment (yrao6597) [reply Backend/app/controllers/volunteer.controller.js]: Thanks for the reminder! On it now

File: Backend/app/controllers/partner.controller.js [hunk 10]
```diff
@@ -38,6 +39,20 @@ const partnerController = {
...
+    try {
+      logger.debug(req.body);
+      const new_partner = new Partner(req.body);
+      const partner_id = await new_partner.create();
+      logger.debug(partner_id);
+      res.json(partner_id);
+    } catch (err) {
+      next(err);
+    }
+>>>>>>> origin/main
```
Comment (youssefsoli) [inline Backend/app/controllers/partner.controller.js:55]: Incomplete merge

File: Backend/app/controllers/supporter.controller.js [hunk 11]
```diff
@@ -12,12 +12,20 @@ const supporterController = {
...
     try {
+<<<<<<< HEAD
       const new_supporter = new Supporter(req.body);
       const supporter_id = await new_supporter.create();
+=======
+      logger.debug(req.body);
+      const new_supporter = new Supporter(req.body);
+      const supporter_id = await new_supporter.create();
+      logger.debug(supporter_id);
+>>>>>>> origin/main
```
Comment (youssefsoli) [inline Backend/app/controllers/supporter.controller.js:23]: ditto

File: Backend/app/controllers/supporter.controller.js [hunk 12]
```diff
@@ -12,12 +12,20 @@ const supporterController = {
...
       res.json(supporter_id);
     } catch (err) {
       next(err);
     }
+<<<<<<< HEAD
```
Comment (youssefsoli) [inline Backend/app/controllers/supporter.controller.js:28]: ditto

File: Backend/app/controllers/volunteer.controller.js [hunk 13]
```diff
@@ -10,6 +10,7 @@ const volunteerController = {
       next(err);
     }
   },
+<<<<<<< HEAD
```
Comment (youssefsoli) [inline Backend/app/controllers/volunteer.controller.js:13]: ditto

File: Backend/app/controllers/volunteer.controller.js [hunk 14]
```diff
@@ -39,6 +40,16 @@ const volunteerController = {
...
+=======
+
+  async create(req, res, next) {
+    try {
+      logger.debug(req.body);
+      const new_volunteer = new Volunteer(req.body);
+      const volunteer_id = await new_volunteer.create();
+      logger.debug(volunteer_id);
+      res.json(volunteer_id);
+>>>>>>> origin/main
```
Comment (youssefsoli) [inline Backend/app/controllers/volunteer.controller.js:52]: ditto

**Review by youssefsoli — CHANGES_REQUESTED** [review]

**Pull request closed without merging**